File: Linked_List_Cycle_II/solution.best.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def detectCycle(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        
        if not head: return head
        
        slow = head.next
        fast = head.next
        if fast != None: fast = fast.next
        while slow != fast and fast != None:
          slow = slow.next
          fast = fast.next
          if fast != None: fast = fast.next
        
        if not fast: return fast
        
        slow2 = head
        while slow != slow2:
          slow = slow.next
          slow2 = slow2.next
        return slow
        
        
        
        

File: Linked_List_Cycle_II/solution.best.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
  
  public ListNode detectCycle(ListNode head) {
    if (head == null) return null;
    
    ListNode slow = head;
    ListNode fast = head;
    do {
      slow = slow.next;
      fast = fast.next;
      if (fast != null) fast = fast.next;
    } while (slow != fast && fast != null);
    
    if (fast == null) return null;
      
    ListNode slow2 = head;
    while (slow2 != slow) {
      slow = slow.next;
      slow2 = slow2.next;
    }
    return slow;
  }
}
